server/py/services/api/crud/pipelines.py

```diff
@@ -315,6 +273,72 @@ def create_pipeline(
     def initialize_kfp_client(namespace: typing.Optional[str] = None):
         return mlrun_pipelines.utils.get_client(mlrun.mlconf.kfp_url, namespace)
 
+    def _paginate_runs(
+        self,
+        kfp_client: mlrun_pipelines.utils.kfp.Client,
+        page_token: typing.Optional[str] = None,
+        page_size: typing.Optional[int] = None,
+        sort_by: typing.Optional[str] = None,
+        filter_: typing.Optional[str] = None,
+    ) -> tuple[int, typing.Optional[int], list[mlrun_pipelines.models.PipelineRun]]:
+        next_page_token = -1
+        if page_token or page_size:
+            runs, next_page_token = self._list_runs_from_kfp(
+                kfp_client,
+                page_token,
+                page_size or mlrun.common.schemas.PipelinesPagination.default_page_size,
+                sort_by,
+                filter_,
+            )
+            runs = [mlrun_pipelines.models.PipelineRun(run) for run in runs or []]
+            # In-memory filtering turns Kubeflow's counting inaccurate if there are multiple pages of data
+            # so don't pass it to the client in such case
+            total_size = -1 if next_page_token else len(runs)
+        else:
+            runs = []
+            while next_page_token:
+                page_runs, next_page_token = self._list_runs_from_kfp(
+                    kfp_client,
+                    page_token,
+                    page_size or mlrun.common.schemas.PipelinesPagination.max_page_size,
+                    sort_by,
+                    filter_,
+                )
+                runs.extend(
+                    [mlrun_pipelines.models.PipelineRun(run) for run in page_runs or []]
+                )
```

Comment:
Maybe move this section into a separate function that calls `_list_runs_from_kfp` and directly returns a list of `PipelineRun` instances, and it would help to add comments before each if and else block to clarify the purpose.